### PR TITLE
[WIP] Build custom Redis client based on the installed extension

### PR DESCRIPTION
**Pull request by phansys**
|Q            |A     |
|---          |---   |
|Branch       |master|
|Bug fix?     |no    |
|New feature? |yes   |
|BC breaks?   |no    |
|Deprecations?|no    |
|Tests pass?  |yes   |
|Fixed tickets|#399  |
|License      |MIT   |
|Doc PR       |n/a   |

### CODE REVIEW

**Comment by phansys** [comment]
I'm trying to use the reflection API in order to introspect the `\Redis` class, but it seems to fail to get some method arguments. By instance, with `pconnect()` it doesn't find any argument. Do you know how could I get a valid representation of the class provided by the extension?
Thank you in advance.

**Comment by B-Galati** [comment]
See #442 if you want more details ;-)

File: .travis.yml [hunk 1]
```diff
@@ -56,4 +59,6 @@ before_install:
 
 install: composer update
 
-before_script: mkdir Snc && ln -s ../ Snc/RedisBundle
+before_script:
+  - mkdir Snc && ln -s ../ Snc/RedisBundle
+  - php clientbuilder/build_redis_client.php
```
Comment (B-Galati) [inline .travis.yml]: I am not sure why it is necessary to have that in the CI.

IMHO, the class should be generated at cache warmup then dump into var/cache/snc-redis-bundle and then loaded thanks to a custom autoloader. It's basically what doctrine-bundle is doing for entities.

What do you think?

Comment (phansys) [reply .travis.yml]: I think that could work only when using DI, but be aware that `PhpredisClientFactory` doesn't depend on the container.
Also, in my first approach I was thinking to check the generated file in each build in order to detect any change based on the installed Redis version, but I think it doesn't have sense now that the installed versions are set explicitly at Travis config.

**Comment by phansys** [comment]
Thank you for pointing me in the right direction. I've opened phpredis/phpredis#1526 for the described problem.

**Comment by B-Galati** [comment]
@phansys Given what was said in the issue, I think we could only support phpredis4 for this feature and have a fallback compatible with phpredis3 only. That's quite easy to do since v3 will not change anymore.
We could provide another one for phpredis2 but it's not important IMHO as it's quite an old version.

**Comment by phansys** [comment]
Sure. I'll try to work on the proposed fix today. Thank you.

**Comment by phansys** [comment]
Methods like `echo()` and `eval()` are present at Redis 3.1.6, but these are not allowed by PHP < 7.0. Should we remove them from the custom client based on that version or should we restrict the usage of the client only for PHP >= 7?
For BC reasons, I think the first option is more suitable for this case, but I'd like to hear your opinion.

**Comment by B-Galati** [comment]
> By the other hand, Redis `4.3.0` seems to be not compatible with `4.2.0` ([ref](https://travis-ci.org/snc/SncRedisBundle/jobs/508445757)):
> 
> ```
> Declaration of Snc\RedisBundle\Client\Phpredis\Client4_3_0::psubscribe(array $patterns, $callback) should be compatible with Redis::psubscribe(array $patterns)
> ```
> 
> Should we ship a client by minor version?

For phpredis4 we could only generate the client class on cache warmup so that it would always fit with the installed phpredis version.

We would have static client class only for older phpredis version, (e.g.) v2 and v3.


> Methods like `echo()` and `eval()` are present at Redis 3.1.6, but these are not allowed by PHP < 7.0. Should we remove them from the custom client based on that version or should we restrict the usage of the client only for PHP >= 7?
> For BC reasons, I think the first option is more suitable for this case, but I'd like to hear your opinion.

For v3, the bundle only accepts PHP 7 so let's say we add this feature for V3 only. What do you think?
Which should be released in less than 2 months (I hope).

**Comment by phansys** [comment]
I think your proposal is doable. Regarding the release of `v3.0.0`, should this PR target `master`?

**Comment by B-Galati** [comment]
Yes `master` for upcoming V3.

**Comment by B-Galati** [comment]
Should we close this PR ?

**Comment by phansys** [comment]
>Should we close this PR ?
Yes, closing since this fix will not be accepted as is for `2.x` branch.

**Pull request closed without merging**